src/callbacks_step/save_solution.jl

```diff
@@ -52,25 +63,64 @@ function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{<:Any, <:Sav
   end
 end
 
+function Base.show(io::IO, ::MIME"text/plain",
+                   cb::DiscreteCallback{<:Any, <:PeriodicCallbackAffect{<:SaveSolutionCallback}})
+  @nospecialize cb # reduce precompilation time
 
-function SaveSolutionCallback(; interval=0,
+  if get(io, :compact, false)
+    show(io, cb)
+  else
+    save_solution_callback = cb.affect!.affect!
+
+    setup = [
+             "dt" => save_solution_callback.interval,
+             "solution variables" => save_solution_callback.solution_variables,
+             "save initial solution" => save_solution_callback.save_initial_solution ? "yes" : "no",
+             "save final solution" => save_solution_callback.save_final_solution ? "yes" : "no",
+             "output directory" => abspath(normpath(save_solution_callback.output_directory)),
+            ]
+    summary_box(io, "SaveSolutionCallback", setup)
+  end
+end
+
+
+function SaveSolutionCallback(; interval::Integer=0,
+                                dt=0.0,
                                 save_initial_solution=true,
                                 save_final_solution=true,
                                 output_directory="out",
                                 solution_variables=cons2prim)
 
-  solution_callback = SaveSolutionCallback(interval, save_initial_solution, save_final_solution,
+  if dt > 0
+    interval = dt
+  end
+
+  solution_callback = SaveSolutionCallback(interval,
+                                           save_initial_solution, save_final_solution,
                                            output_directory, solution_variables)
 
-  DiscreteCallback(solution_callback, solution_callback, # the first one is the condition, the second the affect!
-                   save_positions=(false,false),
-                   initialize=initialize!)
+  if dt > 0
+    # Add a `tstop` every `dt`, and save the final solution.
+    return PeriodicCallback(solution_callback, dt,
+                            initialize=initialize_save_cb!,
+                            final_affect=save_final_solution)
```

Comment:
```suggestion
    return PeriodicCallback(solution_callback, dt,
                            save_positions=(false, false),
                            initialize=initialize_save_cb!,
                            final_affect=save_final_solution)
```
It can pass all keyword arguments to `DiscreteCallback`, doesn't it? If it's using the same defaults, we need to add this to avoid storing stuff in RAM.

Reply:
Did you run some tests with this, checking whether it starts to allocate more than it should?

Reply:
Did you run some tests with this, checking whether it starts to allocate more than it should?

Reply:
I have not run any tests on this yet to examine allocations. Would I do this with BenchmarkTools? Or is it enough to run a few elixirs and examine the behaviour?

Reply:
Yes - I would just run some not-too-short simulation with the new and old settings and compare the memory usage.

Reply:
Done. I now pass the same defaults.

Reply:
> Yes - I would just run some not-too-short simulation with the new and old settings and compare the memory usage.

I picked two elixirs to try out and compare the memory usage. They are nearly identical, which is good news. I used one elixir with a static mesh and one with AMR to double check that nothing funny was going on.

1. `tree_2d_dgsem/elixir_euler_ec.jl` with output intervals taken so the same number of files are created.
```julia
# this PR with `dt=0.1`
 I/O                            6   21.9ms    7.6%  3.64ms   7.67MiB   99.4%  1.28MiB
   save solution                5   19.4ms    6.7%  3.88ms   7.57MiB   98.1%  1.51MiB
   ~I/O~                        6   2.41ms    0.8%   402μs   89.3KiB    1.1%  14.9KiB
   get element variables        5   59.3μs    0.0%  11.9μs   13.0KiB    0.2%  2.61KiB
   save mesh                    5    775ns    0.0%   155ns     0.00B    0.0%    0.00B

# `main` branch with `interval=10`
 I/O                            6   20.7ms    7.5%  3.45ms   7.67MiB   99.4%  1.28MiB
   save solution                5   18.0ms    6.5%  3.60ms   7.57MiB   98.2%  1.51MiB
   ~I/O~                        6   2.57ms    0.9%   428μs   89.3KiB    1.1%  14.9KiB
   get element variables        5    109μs    0.0%  21.8μs   10.3KiB    0.1%  2.06KiB
   save mesh                    5    923ns    0.0%   185ns     0.00B    0.0%    0.00B
```
2. `tree_2d_dgsem/elixir_advection_amr.jl` with output intervals taken so the same number of files are created.
```julia
# this PR with `dt=1.0`
 I/O                              12   52.0ms   33.5%  4.33ms   2.30MiB   25.8%   196KiB
   save solution                  11   25.2ms   16.3%  2.29ms   1.77MiB   19.9%   165KiB
   save mesh                      11   22.6ms   14.6%  2.05ms    427KiB    4.7%  38.9KiB
   ~I/O~                          12   3.73ms    2.4%   311μs   38.3KiB    0.4%  3.20KiB
   get element variables          11    459μs    0.3%  41.8μs   79.8KiB    0.9%  7.25KiB

# `main` branch with `interval=15`
 I/O                              12   54.2ms   37.6%  4.52ms   2.30MiB   26.3%   196KiB
   save solution                  11   25.1ms   17.4%  2.28ms   1.77MiB   20.3%   165KiB
   save mesh                      11   24.8ms   17.2%  2.25ms    429KiB    4.8%  39.0KiB
   ~I/O~                          12   3.94ms    2.7%   328μs   38.3KiB    0.4%  3.20KiB
   get element variables          11    403μs    0.3%  36.6μs   70.0KiB    0.8%  6.36KiB
```

Reply:
LGTM! These are the times from a subsequent run, i.e., without compilation, right?

Reply:
Sorry, I wasn't clear enough. I didn't mean memory usage of the stuff we monitor with our timer but the memory usage of OrdinaryDiffEq.jl potentially saving intermediate solutions. After a full simulation, we would like to have `length(sol.u) == length(sol.t) == 2`. Without the `save_positions` setting I mentioned above, I would expect OrdinaryDiffEq.jl to save a full solution vector in RAM whenever the new callback is activated.

Reply:
You can of course also monitor the memory usage of the Julia process, @andrewwinters5000 - that's what I had in mind initially.

Reply:
> LGTM! These are the times from a subsequent run, i.e., without compilation, right?

Yes, sorry. i forgot to mention that

Reply:
Do you still have full output from these runs? You can have a look at the `alloc'c memory` output in the analysis callback. If OrdinaryDiffEq.jl stores additional stuff, it should show up there as well.

Reply:
> You can of course also monitor the memory usage of the Julia process, @andrewwinters5000 - that's what I had in mind initially.

Ah, okay. I misunderstood what you mean. I can investigate this usage by OrdinaryDiffEq as well.

Reply:
Tracking the memory usage in Activity Monitor on Mac there is a slight increase in the memory usage when I run `tree_2d_dgsem/elixir_advection_amr.jl`. The `interval` strategy uses 1.84Gb whereas the `dt` strategy uses 1.86Gb. But this number changes a bit from run to run, so I am not sure how reliable it is (or if I am checking this correctly). Using Michael's suggestion, the allocated memory for the analysis callback are nearly identical

- `interval` strategy
```julia
 analyze solution                  3   10.1ms    6.0%  3.38ms   50.7KiB    0.5%  16.9KiB
```
- `dt` strategy
```julia
analyze solution                  3   10.3ms    6.9%  3.44ms   50.8KiB    0.6%  16.9KiB
```

Reply:
I can confirm that with the `save_positions` setting we have that `length(sol.u) == length(sol.t) == 2` after trying out the two elixirs mentioned above in this discussion. If I deactivate the `save_positions` then for, e.g., the AMR elixir I get that `length(sol.u)` is 22, which we avoid.

Reply:
Perfect, thank you for also checking that the behavior is as expected. I'm resolving this convo now.